### PR TITLE
cider-doc now renders spec using `cider-browse-spec--pprint-indented`

### PR DESCRIPTION
**Pull request by jpmonettas**
There is also a button to jump to the browser

**This needs to be merged together with!** https://github.com/clojure-emacs/cider-nrepl/pull/472

-----------------

- [X] The commits are consistent with our [contribution guidelines][1]
- [ ] You've added tests (if possible) to cover your change(s)
- [ ] All tests are passing (`make test`)
- [ ] The new code is not generating bytecode or `M-x checkdoc` warnings
- [ ] You've updated the changelog (if adding/changing user-visible functionality)
- [ ] You've updated the readme (if adding/changing user-visible functionality)
- [ ] You've updated the refcard (if you made changes to the commands listed there)

### CODE REVIEW

**Comment by jpmonettas** [comment]
@bbatsov **This is not ready to be merged yet**, it works with eval-buffer but has a compilation error due to dependency cycle. I'm creating the pull request to discuss the solution.

The thing is I need to create a button in cider-doc.el that when clicked calls a procedure inside cider-browse-spec.el.

Can someone with more experience in elisp help me with this? Thanks

**Comment by xiongtx** [comment]
A screenshot would be helpful:

image

This presentation could be improved. See the [existing](https://github.com/clojure-emacs/cider/issues/2150#issue-285540638):

image

**Comment by jpmonettas** [comment]
@xiongtx Don't know why it looks like that, mines looks different :

latest-screenshot

I prefer it to show anonymous functions as functions and not to remove the #, because they look like the spec is invalid.

**Comment by xiongtx** [comment]
Hmm, I'm not sure what's causing the differences. Perhaps @bbatsov can see what results he gets.

**Comment by bbatsov** [comment]
Perhaps I'm missing something, but the first version does look a bit tidier to me. Perhaps we can make it configurable which version to display in the doc buffer? After all I guess the "Browse" button is the most important thing.

**Comment by xiongtx** [comment]
Which version looks tidier? Couldn't be the first one in [my post](https://github.com/clojure-emacs/cider/pull/2154#issuecomment-355682134)--it doesn't even line up the `args`, `ret`, etc.

Anyways, I'd be fine w/ merging what @jpmonettas [shows](https://github.com/clojure-emacs/cider/pull/2154#issuecomment-355685909), though I don't think the `s/fspec` is necessary--it's just noise. That it's an `fdef` is obvious from the `:args`, `:ret`, etc.

File: cider-browse-spec.el [hunk 1]
```diff
@@ -243,6 +245,14 @@ Display TITLE at the top and SPECS are indented underneath."
                        (string-join (mapcar #'cider-browse-spec--pprint (cl-rest form)) " "))))))
         (t (format "%s" form))))
 
+(defun cider-browse-spec--pprint-indented (spec-form)
```
Comment (xiongtx) [inline cider-browse-spec.el]: Seems this can be simplified w/ `cider-font-lock-as-clojure` on the pprinted string.

**Comment by bbatsov** [comment]
By first, I meant the existing one. I like the nice structural distinction between `args`, `ret` and `fn` (and I also think we can use better labels for those, btw). My personal preference would be to show by default something easier to parse by human minds and optionally be able to show the actual spec as it exists in the code.

**Comment by jpmonettas** [comment]
@bbatsov @xiongtx Just removed the surrounding (s/fspec ...) stuff, now it looks like:

latest-screenshot

The cat looks nicer in our current version because it only has two parts, so we can show it all in one line, but as soon as you have a larger cat line I think it looks nicer to print one part in each line.

> My personal preference would be to show by default something easier to parse by human minds and optionally be able to show the actual spec as it exists in the code.

I think this too, that's why I created that elisp spec formatter, so we have control and can fine tune how we print a spec instead of relaying on vanilla s/form pprinting.

Some time ago I created [pretty-spec](https://github.com/jpmonettas/pretty-spec) that uses [fipp](https://github.com/brandonbloom/fipp) to pretty print specs sensitive to a width you have defined. Using something like this will require to move pprinting from elisp to cider-nrepl. (I explored those ideas some time ago when writing [Inspectable](https://github.com/jpmonettas/inspectable)

File: cider-doc.el [hunk 2]
```diff
@@ -461,18 +461,14 @@ Tables are marked to be ignored by line wrap."
...
-              (insert (format "%-4s: " role))
-              (thread-first desc
-                cider-sync-request:format-code
-                cider-font-lock-as-clojure
-                (split-string "\n")
-                insert-rectangle))
-            (insert "\n"))
-          (insert "\n"))
+          (emit "Spec:" 'font-lock-function-name-face)
+          (insert (cider-browse-spec--pprint-indented spec))
```
Comment (xiongtx) [inline cider-doc.el]: You should probably `(require 'cider-browse-spec)` at the top of the file.

**Comment by xiongtx** [comment]
Yes, removing the `s/fspec` stuff cleans it up a lot. I'm good w/ how it looks now.

**Comment by jpmonettas** [comment]
Maybe this looks better? What about the naming ?

latest-screenshot

File: cider-doc.el (hunk 2 above)
Comment (xiongtx) [reply cider-doc.el]: Try to figure out what's causing the dependency cycle then, and break it. We probably have some unnecessary `require`s in some files, or maybe some functions need to be shuffled around. 

In any case, you're using `cider-browse-spec--...` in `cider-doc.el` so you should definitely `require` it in `cider-doc.el`.

Comment (jpmonettas) [reply cider-doc.el]: Yeah, was cider-browse-spec depending on cider-interaction, it doesn't need to, so this is fixed.

**Comment by jpmonettas** [comment]
@bbatsov @xiongtx if both agree on the latest screenshot now everything is ready to be merged together with https://github.com/clojure-emacs/cider-nrepl/pull/472

- [X] The commits are consistent with our [contribution guidelines][1]
- [ ] You've added tests (if possible) to cover your change(s)
- [X] All tests are passing (`make test`)
- [X] The new code is not generating bytecode or `M-x checkdoc` warnings
- [ ] You've updated the changelog (if adding/changing user-visible functionality)
- [ ] You've updated the readme (if adding/changing user-visible functionality)
- [ ] You've updated the refcard (if you made changes to the commands listed there)

**Comment by bbatsov** [comment]
I'm fine with it. Just rebase on top of the current `master`.

**Comment by bbatsov** [comment]
You should also add some changelog entry.

**Comment by jpmonettas** [comment]
@bbatsov added an entry to the changelog and rebased on top of master.

File: cider-doc.el [hunk 3]
```diff
@@ -461,18 +462,14 @@ Tables are marked to be ignored by line wrap."
...
+          (insert "\n\n")
+          (insert-text-button "Browse spec"
+                              'follow-link t
+                              'action (lambda (x)
+                                        (cider-browse-spec (format "%s/%s" ns name))))
```
Comment (bbatsov) [inline cider-doc.el]: I noticed at the build output `x` is unused, which I assume is a mistake. It should be just `_` if it's not needed.

**Pull request merged**